Allow External Urls / Parse Remote Markdown files

kzap opened the pull request:
With this update you can link to absolute URLs in your Table of Contents and it will link to it without trying to parse it.

More importantly, if you have a markdown file somewhere else on the internet or in another GIthub repository you can use it in your Table of Contents, you can even include `username:password` or `token` in the URL for private repos. Also works with Bitbucket URLs:

<img width="1435" alt="screenshot 2017-08-13 14 57 46" src="https://user-images.githubusercontent.com/62213/29246885-f3ae6f70-8037-11e7-8cd4-eabb8d96f4fa.png">

Sample Table of Contents `docs/README.md`
```
* [docpress](../README.md)
* [External Link to Github](https://github.com)
* [Include Remote Github Markdown]
(https://github.com/docpress/docpress/blob/master/README.md)
* [Include Another Remote Github Markdown](https://github.com/docpress/docpress/blob/master/docs/showcase.md)
* [Include Remote Bitbucket Markdown](https://bitbucket.org/MartinFelis/love-android-sdl2/src/a1aca69aa453ad41f45234c8d7739136637acc72/README.md)
```

knownasilya commented on lib/fix_html.js:
```diff
@@ -56,17 +57,37 @@ function fixReferences ($, fname, sources, files, page) {
+
+      let baseSource = parsedUrl.href
+      baseSource = baseSource.split('/')
+      baseSource.splice(-1)
```
can we use `slice` instead, and assign the value.

kzap replied:
I changed this to use dirname() instead

kzap commented:
Improved the error checking, fixed linter errors and also made remote markdown files link to each other if they are both in the Table of Contents. Example:
```
* [Include Remote Github Markdown]
(https://github.com/docpress/docpress/blob/master/README.md)
* [Include Another Remote Github Markdown](https://github.com/docpress/docpress/blob/master/docs/showcase.md)```

knownasilya commented on lib/fix_html.js:
```diff
@@ -56,17 +58,60 @@ function fixReferences ($, fname, sources, files, page) {
-    if (!sources[target]) {
-      throw new Error(`${base}: Unknown reference '${origUrl}'`)
+    // if absolute markdown file, make link relative to MD file
+    if (page.absolute) {
```
Currently this isn't tested. If you are still interested, do you think you could add a test? I rebased your branch and updated the tests to pass, but this brings down the coverage too much.

kzap commented:
Thank you for updating this but currently I dont have the time to commit to
updating this.

On Wed, 7 Nov 2018 at 3:11 AM Ilya Radchenko <notifications@github.com>
wrote:

> *@knownasilya* requested changes on this pull request.
>
> Need to add a few tests to cover the new code, see
> https://coveralls.io/builds/19940891
> ------------------------------
>
> In lib/fix_html.js
> <https://github.com/docpress/docpress-core/pull/196#discussion_r231252823>
> :
>
> > @@ -56,17 +58,60 @@ function fixReferences ($, fname, sources, files, page) {
>        anchor = m[2]
>      }
>
> -    // Get the target source file it points to (eg, `docs/usage.md`).
> -    var target = getTarget(origUrl, base)
> -
> -    // Ensure that it's available.
> -    if (!sources[target]) {
> -      throw new Error(`${base}: Unknown reference '${origUrl}'`)
> +    // if absolute markdown file, make link relative to MD file
> +    if (page.absolute) {
>
> Currently this isn't tested. If you are still interested, do you think you
> could add a test? I rebased your branch and updated the tests to pass, but
> this brings down the coverage too much.
>
> —
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/docpress/docpress-core/pull/196#pullrequestreview-172183474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADzBTn_A82ox4NZIF5N6t67JJ3mkMvMks5usd6-gaJpZM4O1eaz>
> .
>

knownasilya commented:
@rstacruz I'll add the tests 👍
